crates/astria-sequencer-relayer/src/relayer.rs

```diff
@@ -97,62 +92,85 @@ impl Relayer {
         self.state_tx.subscribe()
     }
 
-    async fn get_and_submit_latest_block(&self) -> eyre::Result<State> {
-        let mut new_state = (*self.state_tx.borrow()).clone();
-        let resp = self.sequencer_client.get_latest_block().await?;
-
-        let height = resp.block.header.height.value();
-        if height <= *new_state.current_sequencer_height.get_or_insert(height) {
-            return Ok(new_state);
-        }
-
-        info!(height = ?height, tx_count = resp.block.data.len(), "got block from sequencer");
-        new_state.current_sequencer_height.replace(height);
-
-        if resp.block.header.proposer_address.as_ref() != self.validator.address.as_ref() {
-            let proposer_address = resp.block.header.proposer_address;
-            info!(
-                %proposer_address,
-                validator_address = %self.validator.address,
-                "ignoring block: proposer address is not ours",
+    fn convert_block_response_to_sequencer_block_data(
+        &self,
+        res: BlockResponse,
+        current_height: Option<u64>,
+    ) -> eyre::Result<SequencerBlockData> {
+        if let Some(current_height) = current_height {
+            ensure!(
+                res.block.header.height.value() > current_height,
+                "sequencer block response had height below current height tracked in relayer"
             );
-            return Ok(new_state);
         }
+        ensure!(
+            res.block.header.proposer_address == self.validator.address,
+            "proposer recorded in sequencer block does not match internal validator"
+        );
+        let sequencer_block_data = SequencerBlockData::from_tendermint_block(res.block)
+            .wrap_err("failed converting sequencer block response to sequencer block data")?;
+        Ok(sequencer_block_data)
+    }
 
-        let sequencer_block = match SequencerBlockData::from_tendermint_block(resp.block) {
-            Ok(block) => block,
-            Err(e) => {
-                warn!(error = ?e, "failed to convert block to DA block");
-                return Ok(new_state);
+    #[instrument(skip_all)]
+    async fn submit_blocks(&self, block_responses: Vec<BlockResponse>) -> State {
+        let mut new_state = (*self.state_tx.borrow()).clone();
+        let mut all_converted_blocks = Vec::with_capacity(block_responses.len());
+        for res in block_responses {
+            match self.convert_block_response_to_sequencer_block_data(
+                res,
+                new_state.current_sequencer_height,
+            ) {
+                Ok(converted) => {
+                    if let Err(error) = self.block_tx.send(converted.clone()) {
+                        warn!(?error, "failed sending sequencer block data to gossip task");
+                    }
```

Comment:
this will still block on writing to celestia right? this needs to be separated out so that it's gosisped as soon as a new block is received

Reply:
Yeah, you are right. This PR splits reading-from-sequencer and submitting-blocks (which is writing-to-celestia + gossiping-blocks).

The moment we start draining the channel and processing each block we start gossiping, but once we submit to celestia gossiping is blocked again.

I will separate that.

Reply:
I have addressed this in a new commit. The following tasks are now split split such that they will not block one another:

1. receipt of new blocks from the poller (spawn new blocking task to convert to sequencer block data)
2. conversion of block responses to sequencer block data (runs in blocking tasks on a joinset)
3. queuing + gossip each block data
4. submitting the queued block data to celestia in a separate task.